### DISCUSSION TITLE
The `pow` function in `UD60x18` does not work with bases lower than 1

### DISCUSSION THREAD

**Discussion by jinqiao0423** [PaulRBerg/prb-math#172] [Q&A] [answered]
Hi! I'm not very sure about the input lower & upper bounds, as well as precision in some scenarios such as x^y. 

My code is something like this:
```
function testUD60x18Pow(uint base, uint exp) external pure returns (uint) {
    UD60x18 convertedBase = ud(base);
    UD60x18 convertedExp = ud(exp);
    return convertedBase.pow(convertedExp).intoUint256();
}
```
From the name I assume that UD60x18 is with a 1e18 precision so an 1e18 input is actually 1. 
I tried 1 ^ 1 = 1, 0.1^1 = 1, 1^0.1 = 1, fine, but in case of 0.1^0.1, an error occurs:
```
call to Playground.testUD60x18Pow errored: Internal JSON-RPC error.{
  "code": 1,
  "message": "message execution failed: exit 33, revert reason: 0x36d32ef0000000000000000000000000000000000000000000000000016345785d8a0000, vm error: message failed with backtrace:\n00: f019287 (method 3844450837) -- contract reverted (33)\n (RetCode=33)"
}
```
I also tried some other cases such as:
1.1^1.1, succeeded
1.000000000000000001^1.000000000000000001, succeeded
10^10, succeeded
100^100, failed
1000^1000,failed
And 0.9^0.9 also failed.. It seems if base and exp are both below 1, tx will revert?
I understand that there must be an upper bounds because of overflow. However, I'm not sure whether there is a lower bounds on UD60x18 in such scenarios. If so, what is the exact lower bound? Thx!

**Comment by PaulRBerg** [comment]
Hi @jinqiao0423, thank you opening this discussion.

This is not a bug. It's a known limitation of the `pow` function. If you run the function you posted on Remix, you will see this error reported when passing `0.1^0.1`:

```text
Error provided by the contract:
PRBMath_UD60x18_Log_InputTooSmall : Emitted when taking the logarithm of a number less than 1.
```

The reason for this limitation is that the [`pow`](https://github.com/PaulRBerg/prb-math/blob/6e1a5b0dc113efada32bb0fc0f989ee0ce970e1a/src/ud60x18/Math.sol#L436) function uses the [`log2`](https://github.com/PaulRBerg/prb-math/blob/6e1a5b0dc113efada32bb0fc0f989ee0ce970e1a/src/ud60x18/Math.sol#L367) function under the hood, which does not work for values lower than `UNIT` (1e18), because the resultant logarithm would be negative. The `UD60x18` type does not allow intermediary negative values.

The solution is simple: just use `SD59x18` instead of `UD60x18`, which can handle negative values. This topic has also been discussed here:

https://github.com/PaulRBerg/prb-math/discussions/50#discussioncomment-1369641

> It seems if base and exp are both below 1, tx will revert?

Actually, the revert happens only when the base is lower than 1 (see the implementation of [`pow`](https://github.com/PaulRBerg/prb-math/blob/6e1a5b0dc113efada32bb0fc0f989ee0ce970e1a/src/ud60x18/Math.sol#L436)).

**Comment by PaulRBerg** [comment, marked as answer]
**Update**: since https://github.com/PaulRBerg/prb-math/pull/182 has been merged, the `UD60x18.pow` function works with bases lower than 1e18 🎉 

Refer to the NatSpec comments for documentation:

https://github.com/PaulRBerg/prb-math/blob/97170d0f52de76edfb5fe63c5810a8705c3b0ea7/src/ud60x18/Math.sol#L453-L481
